homeassistant/components/cups/sensor.py

```diff
@@ -120,21 +151,111 @@ def update(self):
+        return ICON
+
+    @property
+    def available(self):
+        """Return True if entity is available."""
+        return self._available
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        if self._attributes is not None:
+            try:
+                return next(v for k, v in PRINTER_STATES.items()
+                            if self._attributes['printer-state'] == k)
+            except StopIteration:
+                return self._attributes['printer-state']
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        if self._attributes is not None:
+            state_attributes = {}
+
+            if 'printer-info' in self._attributes:
+                state_attributes[ATTR_PRINTER_INFO] = \
+                    self._attributes['printer-info']
+
+            if 'printer-location' in self._attributes:
+                state_attributes[ATTR_PRINTER_LOCATION] = \
+                    self._attributes['printer-location']
+
+            if 'printer-state-message' in self._attributes:
+                state_attributes[ATTR_PRINTER_STATE_MESSAGE] = \
+                    self._attributes['printer-state-message']
+
+            if 'printer-state-reasons' in self._attributes:
+                state_attributes[ATTR_PRINTER_STATE_REASON] = \
+                    self._attributes['printer-state-reasons']
+
+            if 'printer-uri-supported' in self._attributes:
+                state_attributes[ATTR_PRINTER_URI_SUPPORTED] = \
+                    self._attributes['printer-uri-supported']
+
+            return state_attributes
+
+    def update(self):
+        """Fetch new state data for the sensor."""
+        self.data.update()
+        self._attributes = self.data.attributes.get(self._name)
+        self._available = self.data.available
 
 
 # pylint: disable=no-name-in-module
 class CupsData:
     """Get the latest data from CUPS and update the state."""
 
-    def __init__(self, host, port):
+    def __init__(self, host, port, ipp_printers):
         """Initialize the data object."""
         self._host = host
         self._port = port
+        self._ipp_printers = ipp_printers
+        self.is_cups = (ipp_printers is None)
         self.printers = None
+        self.attributes = {}
+        self.available = False
 
     def update(self):
         """Get the latest data from CUPS."""
         cups = importlib.import_module('cups')
 
-        conn = cups.Connection(host=self._host, port=self._port)
-        self.printers = conn.getPrinters()
+        try:
+            conn = cups.Connection(host=self._host, port=self._port)
+            if self.is_cups:
+                self.printers = conn.getPrinters()
+            else:
+                for ipp_printer in self._ipp_printers:
+                    self.attributes[ipp_printer] = conn.getPrinterAttributes(
+                        uri="ipp://" + self._host
```

Comment:
Please use string formatting instead of concatenation: `"ipp://{}:{}/{}".format(self._host, self._port, ipp_printer)`